slow_query_exporter.rb

```diff
@@ -0,0 +1,230 @@
+#!/usr/bin/env ruby
+
+require 'gelf'
+require 'digest/crc64'
+require 'optparse'
+
+class QueryParser
+  def initialize
+    @buffer = ""
+    @parsed = []
+    @query = SlowQuery.new
+  end
+
+  def parse_line(line)
+    @query.parse_line(line)
+    if @query.done?
+      @parsed << @query
+      @query = SlowQuery.new
+    end
+  end
+
+  def pop_finished_query
+    @parsed.pop
+  end
+end
+
+class SlowQuery
+  SPECIAL_PREFIXES = {
+    query_string: "_",
+    request_time: "_",
+    host: "_",
+    timestamp: "",
+  }
+  PATTERNS = {
+    header:   /(^Tcp port:|^Time\s+Id|started with:$)/,
+    skip:     /^(# Time: \d+|use \w+;)/,
+    userhost: /^# User@Host: (\S+)\[\S+\] @\s+\[(\S+)\]/,
+    thread:   /^# Thread_id: (\d+)\s+Schema: (\w+)\s+Last_errno: (\d+)\s+Killed: (\d+)/,
+    qtime:    /^# Query_time: (\S+)\s+Lock_time: (\S+)\s+Rows_sent: (\d+)\s+Rows_examined: (\d+)\s+Rows_affected: (\d+)\s+Rows_read: (\d+)/,
+    bytes:    /^# Bytes_sent: (\d+)\s+Tmp_tables: (\d+)\s+Tmp_disk_tables: (\d+)\s+Tmp_table_sizes: (\d+)/,
+    trxid:    /^# InnoDB_trx_id: (\S+)/,
+    qchit:    /^# QC_Hit: (\w+)\s+Full_scan: (\w+)\s+Full_join: (\w+)/,
+    filesort: /^# Filesort: (\w+)\s+Filesort_on_disk: (\w+)\s+Merge_passes: (\d+)/,
+    innoio:   /^#\s+InnoDB_IO_r_ops: (\d+)\s+InnoDB_IO_r_bytes: (\d+)\s+InnoDB_IO_r_wait: (\S+)/,
+    innowait: /^#\s+InnoDB_rec_lock_wait: (\S+)\s+InnoDB_queue_wait: (\S+)/,
+    innopage: /^#\s+InnoDB_pages_distinct: (\d+)/,
+    time:     /^SET timestamp=(\d+)/,
+    query:    /^(SELECT|INSERT|UPDATE|DELETE)\b/,
+  }
+
+  attr_accessor :attributes
+
+  def initialize
+    @attributes = {query_string: ""}
+    @done = false
+  end
+
+  def parse_line(line)
+    return if line =~ PATTERNS[:header] || line =~ PATTERNS[:skip]
+
+    case line
```

Comment:
Is there any clean way to ensure that these attributes are set correctly and maybe output and error that contains an unparsed input in the chance that any of parameters are missing?

I guess I'm just concerned in the case where the script to fails and we miss out on logging a result?

Maybe just have something in the script to rescue or catch any errors and log as much info on the error as possible instead?